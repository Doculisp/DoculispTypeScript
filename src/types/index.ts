// Export all types for external consumption
export * from './types.ast';
export * from './types.astDoculisp';
export * from './types.astProject';
export * from './types.containers';
export * from './types.controller';
export * from './types.document';
export * from './types.docWriter';
export * from './types.fileHandler';
export * from './types.filePath';
export * from './types.general';
export * from './types.includeBuilder';
export * from './types.internal';
export * from './types.sringBuilder';
export * from './types.stringWriter';
export * from './types.structure';
export * from './types.textHelpers';
export * from './types.tokens';
export * from './types.trimArray';
export * from './types.variableTable';
export * from './types.version';