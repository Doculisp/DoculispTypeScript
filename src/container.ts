import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};