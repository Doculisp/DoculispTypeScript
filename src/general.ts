import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};