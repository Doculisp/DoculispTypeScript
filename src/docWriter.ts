import { AstBulletStyle, AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "./types.ast";
import { IRegisterable } from "./types.containers";
import { IDocWriter } from "./types.docWriter";
import { IUtil, Result } from "./types.general";

function buildWriter(util: IUtil): IDocWriter {
    function getLabel(style: AstBulletStyle, load: ILoad): string {
        switch (style) {
            case 'bulleted':
            case 'numbered':
            case 'unlabeled':
                return '';
            default:
                return load.sectionLabel;
        }
    }

    function getNumberedBullet(): (label: string, text: string) => string {
        let cnt = 1;
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            const result = `${cnt}. ${lbl}${text}`;
            cnt++;
            return result;
        }

        return bullet;
    }

    function getBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `* ${lbl}${text}`;
        }

        return bullet;
    }

    function getNoBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `${lbl}${text}`;
        }
    }

    function getBulletBuilder(style: AstBulletStyle): (label: string, text: string) => string {
        switch (style) {
            case 'bulleted':
            case 'bulleted-labeled':
                return getBullet();
            case 'numbered':
            case 'numbered-labeled':
                return getNumberedBullet();
            default:
                return getNoBullet();
        }
    }

    function compileToc(astToc: ITableOfContents, loads: ILoad): string {
        if(astToc.bulletStyle === 'no-table') {
            return '';
        }
    }

    function compileContentLocation(loads: ILoad[]): string {
        const parts: string[] = [];

        loads.forEach(load => {
            if(load.document) {
                const section = compileSection(load.document);
                parts.push(section);
            }
        });

        return `${parts.join('\n')}`;
    }

    function compileHeader(astHeader: IHeader): string {
        const padding = '#'.repeat(astHeader.depthCount);
        return `${padding} ${astHeader.text} ${padding}`;
    }

    function compileTitle(astTitle: ITitle): string {
        return astTitle.label;
    }

    function compileWrite(astWrite: IWrite): string {
        return astWrite.value;
    }

    function compilePart(astPart: AstPart) : string {
        astPart.
    }

    function compileSection(astSection: ISectionWriter): string {
        const parts: string[] = [];

        
    }

    function compile(astResult: Result<IAst>): Result<string> {
        if(!astResult.success) {
            return astResult;
        }

        const ast = astResult.value;

        if(ast.section.type === 'ast-empty') {
            return util.ok('');
        }

        throw new Error("Not yet Implemented");
    }

    function write(path: string, astResult: Result<IAst>): Result<string> {
        throw new Error('Not yet Implemented');
    }

    return {
        compile,
        write,
    };
}

const documentWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'docWriter',
    singleton: true,
    dependencies: ['util', 'fileHandler']
};

export {
    documentWriter,
};