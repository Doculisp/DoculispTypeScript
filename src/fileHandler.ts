import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};