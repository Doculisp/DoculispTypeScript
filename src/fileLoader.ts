import { IRegisterable } from "./types.containers";
import { IFileLoader } from "./types.fileLoader";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileLoader {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileLoader',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};