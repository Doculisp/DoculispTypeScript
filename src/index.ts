#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
