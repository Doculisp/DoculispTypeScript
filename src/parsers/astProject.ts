import { AtomAst, CoreAst, IAstEmpty, RootAst } from "../types/types.ast";
import { IProjectDocument, IProjectDocuments, IProjectParser } from "../types/types.astProject";
import { IDictionary, IRegisterable } from "../types/types.containers";
import { IPath, PathConstructor } from "../types/types.filePath";
import { ILocation, IUtil, Result } from "../types/types.general";
import { IInternals, StepParseResult } from "../types/types.internal";
import { ITrimArray } from "../types/types.trimArray";

interface ISource {
    source: IPath,
    location: ILocation,
    type: 'i-source',
};

interface IOutput {
    output: IPath,
    location: ILocation
    type: 'i-output',
}

function buildAstProject(internals: IInternals, util: IUtil, trimArray: ITrimArray, pathConstructor: PathConstructor): IProjectParser {
    function parse(tokenResults: Result<RootAst | IAstEmpty>): Result<IProjectDocuments> {
        const ids: IDictionary<boolean> = {};

        if(!tokenResults.success) {
            return tokenResults;
        }

        if(tokenResults.value.type === 'ast-Empty') {
            const empty: IProjectDocuments = {
                type: 'project-documents',
                documents: [],
                location: util.location(tokenResults.value.location.documentPath, tokenResults.value.location.documentDepth, tokenResults.value.location.documentIndex, 0, 0),
            };

            return util.ok(empty);
        }

       function parseDocuments(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], IProjectDocuments> {
            const ast = input[0] as CoreAst;

            if(ast.type !== 'ast-atom' && ast.type !== 'ast-container') {
                return internals.noResultFound();
            }

            if(ast.value !== 'documents') {
                return internals.noResultFound();
            }

            if(ast.type === 'ast-atom') {
                const empty: IProjectDocuments = {
                    type: 'project-documents',
                    documents: [],
                    location: ast.location,
                };

                return util.ok({
                    type: 'parse result',
                    location: ast.location,
                    subResult: empty,
                    rest: trimArray.trim(1, input),
                });
            }

            const parser = internals.createArrayParser(parseDocument);
            const maybe = parser.parse(ast.subStructure, (ast.subStructure[0] as AtomAst).location);

            if(!maybe.success) {
                return maybe;
            }

            const [result, remaining] = maybe.value;

            if(0 < remaining.remaining.length) {
                const bad = remaining.remaining[0] as CoreAst;
                return util.fail(`Documents block at '${current.documentPath.fullName}' Line: ${current.line}, Char: ${current.char} contains unknown atom of '${bad.value}' at Line: ${bad.location.line}, Char ${bad.location.char}.`, current.documentPath);
            }

            const documents: IProjectDocuments = {
                type: 'project-documents',
                documents: result,
                location: ast.location,
            };

            return util.ok({
                type: 'parse result',
                location: current,
                subResult: documents,
                rest: trimArray.trim(1, input),
            });
        }

        function parseSource(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], ISource> {
            const ast = input[0] as AtomAst;

            if(ast.type !== 'ast-command') {
                return internals.noResultFound();
            }

            if(ast.value !== 'source') {
                return internals.noResultFound();
            }

            let source: ISource = {
                source: pathConstructor(ast.parameter.value),
                location: current,
                type: 'i-source'
            };

            return util.ok({
                type: 'parse result',
                location: current,
                subResult: source,
                rest: trimArray.trim(1, input),
            });
        }

        function parseOutput(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], IOutput> {
            const ast = input[0] as AtomAst;

            if(ast.type !== 'ast-command') {
                return internals.noResultFound();
            }

            if(ast.value !== 'output') {
                return internals.noResultFound();
            }

            const output: IOutput = {
                output: pathConstructor(ast.parameter.value),
                location: current,
                type: 'i-output'
            };

            return util.ok({
                type: 'parse result',
                location: current,
                subResult: output,
                rest: trimArray.trim(1, input),
            });
        }

        function parseDocumentByParts(originalLocation: ILocation, originalAst: CoreAst[] | false, id?: string | undefined): (input: CoreAst[], current: ILocation) => StepParseResult<CoreAst[], IProjectDocument> {
            return function(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], IProjectDocument> {
                const parser = internals.createArrayParser<CoreAst, IOutput | ISource>(parseSource, parseOutput);

                const maybe = parser.parse(input, current);

                if(!maybe.success) {
                    return maybe;
                }

                const [result, remaining] = maybe.value;

                if(result.length === 0) {
                    return internals.noResultFound();
                }

                if(0 < remaining.remaining.length) {
                    const first = remaining.remaining[0] as AtomAst;
                    return util.fail(`Unknown atom at '${current.documentPath.fullName}' Line: ${first.location.line}, Char: ${first.location.char} of '${first.value}'`, current.documentPath);
                }

                const sources = result.filter(r => r.type === 'i-source');
                const outputs = result.filter(r => r.type === 'i-output');

                if(0 === sources.length && 0 === outputs.length) {
                    return internals.noResultFound();
                }

                if(0 === sources.length) {
                    return util.fail(`Document identifier block at '${originalLocation.documentPath}' Line: ${originalLocation.line}, Char: ${originalLocation.char} does not contain a source block.`, originalLocation.documentPath)
                }

                if(0 === outputs.length) {
                    return util.fail(`Document identifier block at '${originalLocation.documentPath}' Line: ${originalLocation.line}, Char: ${originalLocation.char} does not contain a output block.`, originalLocation.documentPath)
                }

                if(1 < sources.length) {
                    const bad = sources[1] as ISource;
                    return util.fail(`Duplicate source block at '${current.documentPath}' Line: ${bad.location.line}, Char: ${bad.location.char}.`, current.documentPath);
                }

                if(1 < outputs.length) {
                    const bad = outputs[1] as IOutput;
                    return util.fail(`Duplicate output block at '${current.documentPath}' Line: ${bad.location.line}, Char: ${bad.location.char}.`, current.documentPath);
                }

                const source = sources[0] as ISource;
                const output = outputs[0] as IOutput;

                const found: IProjectDocument = {
                    type: 'project-document',
                    destinationPath: output.output,
                    sourcePath: source.source,
                    location: originalLocation,
                    id: id,
                };

                return util.ok({
                    type: 'parse result',
                    location: originalLocation,
                    subResult: found,
                    rest: trimArray.trim(2, originalAst ? originalAst : input),
                });
            }
        }

        function parseDocId(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], IProjectDocument> {
            const ast = input[0] as CoreAst;

            if(ast.type !== 'ast-container') {
                return internals.noResultFound();
            }

            const id = ast.value;

            if(ids[id]) {
                return util.fail(`Duplicate id '${id}' at '${current.documentPath.fullName}' Line: ${current.line}, Char: ${current.char}.`, current.documentPath);
            }

            ids[id] = true;

            return parseDocumentByParts(current, input, id)(ast.subStructure, (ast.subStructure[0] as AtomAst).location);
        }

        function parseDocument(input: CoreAst[], current: ILocation): StepParseResult<CoreAst[], IProjectDocument> {
            const ast = input[0] as CoreAst;

            if(ast.type !== 'ast-container') {
                return internals.noResultFound();
            }

            if(ast.value !== 'document') {
                return internals.noResultFound();
            }

            const parseByParts = parseDocumentByParts(current, false)
            const parser = internals.createArrayParser(parseByParts, parseDocId);

            const maybe = parser.parse(ast.subStructure, (ast.subStructure[0] as AtomAst).location);

            if(!maybe.success) {
                return maybe;
            }

            const [result, remaining] = maybe.value;

            if(0 < remaining.remaining.length) {
                const bad = remaining.remaining[0] as CoreAst;
                return util.fail(`Document block at '${current.documentPath}' Line: ${current.line}, Char: ${current.char} contains unknown block '${bad.value}' at Line: ${bad.location.line}, Char: ${bad.location.char}.`, current.documentPath);
            }

            if(result.length === 0) {
                return util.fail(`Document block at '${current.documentPath}' Line: ${current.line}, Char: ${current.char} does not contain a source or output.`, current.documentPath);
            }

            if(1 < result.length) {
                const bad = result[0] as IProjectDocument;
                return util.fail(`Duplicate block at '${current.documentPath.fullName}' Line: ${bad.location.line}, Char: ${bad.location.char}.`, current.documentPath);
            }

            const doc = result[0] as IProjectDocument;

            return util.ok({
                type: 'parse result',
                location: current,
                subResult: doc,
                rest: trimArray.trim(1, input)
            });
        }

        const parser = internals.createArrayParser(parseDocuments);
        const maybe = parser.parse(tokenResults.value.ast, util.location(tokenResults.value.location.documentPath, tokenResults.value.location.documentDepth, tokenResults.value.location.documentIndex, 1, 1));

        if(!maybe.success) {
            return maybe;
        }

        const [result, remaining] = maybe.value;

        if(0 < remaining.remaining.length) {
            const bad = remaining.remaining[0] as CoreAst;
            return util.fail(`Unknown atom at '${bad.location.documentPath.fullName}' Line: ${bad.location.line}, Char: ${bad.location.char} of '${bad.value}'.`, bad.location.documentPath);
        }

        if(0 === result.length) {
            const empty: IProjectDocuments = {
                type: 'project-documents',
                documents: [],
                location: util.location(tokenResults.value.location.documentPath, tokenResults.value.location.documentDepth, tokenResults.value.location.documentIndex, 0, 0),
            };

            return util.ok(empty);
        }

        if(1 < result.length) {
            const bad = result[1] as IProjectDocuments;
            return util.fail(`Project file may only contain a single documents block. Duplicate documents block detected at '${bad.location.documentPath.fullName}' Line: ${bad.location.line}, Char: ${bad.location.char}.`, bad.location.documentPath);
        }

        return util.ok(result[0] as IProjectDocuments);
    }

    return {
        parse,
    };
}

const doculispParser: IRegisterable = {
    builder: (internals: IInternals, util: IUtil, trimArray: ITrimArray, pathConstructor: PathConstructor) => buildAstProject(internals, util, trimArray, pathConstructor),
    name: 'astProjectParse',
    singleton: false,
    dependencies: ['internals', 'util', 'trimArray', 'pathConstructor']
};

export {
    doculispParser,
};
