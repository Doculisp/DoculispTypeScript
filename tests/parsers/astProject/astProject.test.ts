import { container } from "../../../src/container";
import { configure } from "approvals/lib/config";
import { Options } from "approvals/lib/Core/Options";
import { getVerifier } from "../../tools";
import { buildProjectLocation, testable } from "../../testHelpers";
import { IProjectDocuments } from "../../../src/types/types.astProject";
import { IProjectLocation, Result } from "../../../src/types/types.general";

describe('astProject', () => {
    let parser: (text: string, projectLocation: IProjectLocation) => Result<IProjectDocuments>;
    let verifyAsJson: (data: any, options?: Options) => void;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        parser = testable.project.resultBuilder(container);
    });

    it('should handle an empty project file', () => {
        const result = parser('', buildProjectLocation('./test.dlproj', 1, 1));

        verifyAsJson(result);
    });

    it('should handle an empty documents block', () => {
        const result = parser('(documents)', buildProjectLocation('./test.dlproj', 1, 1));

        verifyAsJson(result);
    });

    describe.skip('basic project documents', () => {
        it('should parse a single document', () => {
            const project = `
(documents
    (document
        (source ./myReadme.md)
        (output ./README.md)
    )
)
`;
            const result = parser(project, buildProjectLocation('./myProject.dlproj', 1, 1));

            verifyAsJson(result);
        });
        
        it('should parse a two document', () => {
            const project = `
(documents
    (document
        (source ./myReadme.md)
        (output ./README.md)
    )
    (document
        (source ./howTo.md)
        (output ./contrib.md)
    )
)
`;
            const result = parser(project, buildProjectLocation('./myProject.dlproj', 1, 1));

            verifyAsJson(result);
        });
    });
});