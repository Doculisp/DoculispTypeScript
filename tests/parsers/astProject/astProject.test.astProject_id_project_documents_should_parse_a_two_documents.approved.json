{
  "success": true,
  "value": {
    "documents": [
      {
        "destinationPath": "../README.md",
        "id": "readme",
        "location": {
          "char": 10,
          "documentDepth": 1,
          "documentIndex": 1,
          "documentPath": "/docs.dlproj",
          "line": 4
        },
        "sourcePath": "./_main.dlisp",
        "type": "project-document"
      },
      {
        "destinationPath": "../contrib.md",
        "id": "contrib",
        "location": {
          "char": 10,
          "documentDepth": 1,
          "documentIndex": 1,
          "documentPath": "/docs.dlproj",
          "line": 10
        },
        "sourcePath": "./contrib/_main.dlisp",
        "type": "project-document"
      }
    ],
    "location": {
      "char": 2,
      "documentDepth": 1,
      "documentIndex": 1,
      "documentPath": "/docs.dlproj",
      "line": 2
    },
    "type": "project-documents"
  }
}
